[SYCL][ESIMD] Add test on simd atomic_update with increment and decrement

vasilytric opened the pull request.

vasilytric committed:
[SYCL][ESIMD] Add test on simd atomic_update with increment and decrement

vasilytric committed:
[SYCL][ESIMD] Change TestDescription structure

Now size, stride etc. doesn't provided into TestDescription and
provided into log::fail()

vasilytric committed:
[SYCL][ESIMD] Add helper that lets obtain sycl::nd_range

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,101 @@
+enum class algorithm_to_change { all, ordered_step, non_ordered_step };
+
+template <int N, algorithm_to_change Algorithm>
+std::vector<size_t> get_indexess_to_change() {
```
```suggestion
std::vector<size_t> get_indexes_to_change() {
```

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,101 @@
+}
+
+template <esimd::atomic_op Operator, typename T>
+T get_expected_value(T base_value, int number_teractions) {
```
```suggestion
T get_expected_value(T base_value, int number_of_iteractions) {
```

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,101 @@
+  } else if constexpr (Operator == esimd::atomic_op::inc) {
+    expected = base_value + number_teractions;
+  }
+  return expected;
```
Note that it left uninitialized in case of any `if` condition above doesn't apply.
Maybe it's better just `return` within `if constexpr` here?

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+using namespace esimd_test::api::functional;
+using namespace sycl::ext::intel::esimd;
+
+// Descriptor class for the case of calling bitwise not operator.
```
```suggestion
// Descriptor class for the case of calling atomic_update overload for zero operands
```

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+
+  template <typename DataT, int NumElems, int NumElemsToChange,
+            typename ChangeElemFilterT, atomic_op ChosenOperator>
+  static void simd_function(DataT *input_data, DataT *output_data,
```
```suggestion
  static void esimd_function(DataT *input_data, DataT *output_data,
```
or
```suggestion
  static void call_esimd_function(DataT *input_data, DataT *output_data,
```
to avoid misinterpretation as reference to the `simd<T,N>` class

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+        functions::get_indexess_to_change<NumElems, AlgorithmToChange>();
+    shared_vector<size_t> shared_indexes_to_change(
+        indexes_to_change.begin(), indexes_to_change.end(),
+        shared_allocator<size_t>(queue));
```
Maybe you could just pass allocator to the `get_indexess_to_change` call?

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+
+    queue.submit([&](sycl::handler &cgh) {
+      DataT *shared_init_values_ptr = shared_init_values.data();
+      DataT *result_ptr = shared_result.data();
```
I'm sorry to mention it one more time, but please, be consistent with naming

vasilytric replied:
I don't know what exactly you mean, but I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+      values_sum += values_before_update[i];
+    }
+
+    output_data[work_item_index] = values_sum;
```
Isn't it just a return value for function?

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch replied:
Some `stash` maybe? Looks like we still have `void` return type here.
```suggestion
    return values_sum;
```

vasilytric replied:
now the `call_esimd_function` will return the `values_sum`, I've done it in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/d26eb3f43e38f873abfd2afa4903c65773c689bc

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+                        functions::filters::ChangeAll,
+                        functions::filters::ChangeNothing>::generate();
+  const auto atomic_op_types =
+      value_pack<atomic_op, atomic_op::inc, atomic_op::dec>::generate_unnamed();
```
Feel free to provide `StringMaker` specialization for `esimd::atomic_op` and use something like:
```suggestion
  const auto atomic_op_types =
      value_pack<atomic_op, atomic_op::inc, atomic_op::dec>::generate_named();
```
or
```suggestion
  const auto atomic_op_types =
      atomic_op_pack<atomic_op::inc, atomic_op::dec>::generate_named();
```

You would be able to clarify in logs which operation failed by that.

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/ab63a6d73d8414bb25b22058560e708893f02f35

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+      if (filter(i) == 1) {
+        ++num_changed_elems;
+      }
+    }
```
Looks like we need to add a comment on what is going on here

vasilytric replied:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24#diff-afaea17121d6706af52229dcfb462795dfe86fde5c98f09b836505a741830b6bR125-R127)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+  const auto filter_types =
+      unnamed_type_pack<functions::filters::ChangeByStep,
+                        functions::filters::ChangeAll,
+                        functions::filters::ChangeNothing>::generate();
```
On the upper level it sounds like we don't need to have a filter "change nothing" with different "number elements to change"
**UPD:** Looks like a naming issue after deeper look. What about something like `masks::ChangeNothing` instead? It's definitely fine to have an all-zero mask for different "number elements to change"

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+  const auto algorithm_to_change_types = value_pack<
+      functions::algorithm_to_change, functions::algorithm_to_change::all,
+      functions::algorithm_to_change::ordered_step,
+      functions::algorithm_to_change::non_ordered_step>::generate_unnamed();
```
Sounds like "unordered step" and "ordered step" is something specific for `filters::ChangeByStep`. And there is no comments here or in `functions_atomic_update.hpp` to clarify it.
**UPD**: Looks like a naming issue after a deeper look. What do you think about using `offset::ordered_step` or `offset_logic::ordered_step` instead? We are talking about `atomic_update(p, offset, mask)` after all.

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+      if (i < *next_selected_index) {
+        if (expected != retrieved) {
+          passed = fail_test(i, expected, retrieved, data_type,
+                             "value that should be updated");
```
Maybe
```suggestion
                             "value that should not be updated");
```
Did I get it right?

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,101 @@
+    step = 1;
+  } else if constexpr (Algorithm == algorithm_to_change::ordered_step) {
+    step = 2;
+  }
```
It would be easier to understand the code if this block would be inside `else` block below:
```
if constexpr (Algorithm == algorithm_to_change::non_ordered_step) {
   ...
} else {
   // this block of code
```

For example, we could easily see that `step` is left uninitialized in a certain case and avoid it either like this:
```
size_t step = 1;
if constexpr (Algorithm == algorithm_to_change::ordered_step) {
  step = 2;
}
```
or with `static_assert`

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,101 @@
+    data = std::vector<size_t>{1, 3, 7, 5};
+    for (size_t i = data.size(); i < N; ++i) {
+      data.push_back(2 * i + 1);
+    }
```
Minor suggestion: we could go in reverse order, from bigger to smaller, to enforce "more unordered" order. Still up to you)

vasilytric replied:
I have used the new algorithm, I've implemented it [here](https://github.com/intel/llvm-test-suite/pull/939/commits/28323ba5bd217c0cce8f4f7785c0ebb2cf8cfd87#diff-28ba9c453e6f6e488b2719b183c72b8802b07162628e6fc1445d3fe6d8206cf6L54-L57)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  if constexpr (Algorithm == algorithm_to_change::non_ordered_step) {
+    data = std::vector<size_t>{1, 3, 7, 5};
```
```suggestion
    data = {1, 3, 7, 5};
```

vasilytric replied:
I have used the new algorithm, I've implemented it [here](https://github.com/intel/llvm-test-suite/pull/939/commits/28323ba5bd217c0cce8f4f7785c0ebb2cf8cfd87#diff-28ba9c453e6f6e488b2719b183c72b8802b07162628e6fc1445d3fe6d8206cf6L54-L57)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,101 @@
+  } else if constexpr (Operator == esimd::atomic_op::inc) {
+    base_value = initial_base_value;
+    step = 0;
+  }
```
```suggestion
  T base_value = initial_base_value;
  T step = 0;

  // Will need to update base value and step for some operators
  static_assert((Operator == esimd::atomic_op::inc) || (Operator == esimd::atomic_op::dec), "Unsupported operator");
```
Did I get it right?

vasilytric replied:
yes, I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+    shared_vector<DataT> shared_result(NumElems, result_fill_value, allocator);
+
+    auto indexes_to_change =
+        functions::get_indexess_to_change<NumElems, AlgorithmToChange>();
```
1) So we are getting a list of up to `NumElems` indexes first. But we use up to `NumElemsToChange` actually.
Maybe comment and `static_assert(NumElemsToChange <= NumElems)`?

2) Isn't there possibility to go out of range here? For example, for `algorithm_to_change::non_ordered_step`:
 - let's say we a running for `NumElems == 16`
 - so we would have `16` indexes here: `{1, 3, 7, 5, 9, 11, 13, 15, 17, ..., 31}`
 - so if we use `NumElemsToChange >= 9` we have an out-of-memory access later on.
 And the problem is that actual values returned by `get_indexess_to_change` is implementation detail, so according to the SOLID principles the only assertion we could assume is: `assert(indexes_to_change.back() < NumElems)`. Given all of this couldn't we just use `get_indexess_to_change<NumElemsToChange>`?

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24
The `static_assert` has been added in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/7e034bb2d1ca8b68f6d317623a13203fed2fc18f

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+    ChangeElemFilterT filter{};
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      changed_indexes[i] = filter(i);
+      offset[i] = indexes_to_change[i] * sizeof(DataT);
```
Actually, we define which indexes to change based on both `offset` and `mask` parameter.
Currently it's hard to differentiate the meaning of `indexes_to_change` and `changed_indexes`. We might stick with `filtered_indexes`, but I would suggest to look if we can mention `mask` and `offset` at the upper-level naming (something like "get_offsets" and "namespace masks").

By doing so we would provide an exact expectation on the upper level (even within `.cpp` files) on what `atomic_update` parameter combinations are actually being used.
I mean, we have an `atomic_update(Tx *p, simd<unsigned, N> offset, simd_mask<N> mask)` already, so why not to re-use the terms from it.

vasilytric replied:
I've applied this comment in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/8f6071023226997f11c6ff346027fdfcf14b5c24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,230 @@
+    log::fail(TestDescriptionT(data_type), "Unexpected value at index ", i,
+              ", retrieved: ", retrieved, ", expected: ", expected,
+              ", for verification type: ", verification_type,
+              ", with number elements to change: ", NumElemsToChange);
```
It's up to you, but this one looks like a test case description detail.

vasilytric replied:
maybe, but I suggest use this way now and leave the option to change it in the future

vasilytric committed:
[SYCL][ESIMD] Improve naming

- Update namespace and variables names
- Simplify statements that should return values from functions

vasilytric committed:
[SYCL][ESIMD] Update functions_atomic_update.hpp

- Move one function out of this file because this is not common
  function
- Add comment to functions

vasilytric committed:
[SYCL][ESIMD] Add static assert

vasilytric committed:
[SYCL][ESIMD] Add operator type to log message

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,85 @@
+};
+} // namespace masks
+
+enum class offset { all, ordered_step, non_ordered_step };
```
```suggestion
enum class offset_generation { all, ordered_step, non_ordered_step };
```
I've already regretted for too short name suggestion, sorry) Maybe something like this would be less probable to conflict with local variables, how do you think?

vasilytric replied:
this comment has been applied [here](https://github.com/intel/llvm-test-suite/pull/939/commits/3a3e37f3d8b5c247a3e92f48f51646d4bf27f524#diff-28ba9c453e6f6e488b2719b183c72b8802b07162628e6fc1445d3fe6d8206cf6R43)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,250 @@
+  const auto atomic_op_types =
+      value_pack<atomic_op, atomic_op::inc, atomic_op::dec>::generate_named(
+          " atomic_op::inc", "atomic_op::dec");
+  const auto algorithm_to_change_types =
+      value_pack<functions::offset, functions::offset::all,
```
```suggestion
  const auto algorithm_to_change_offset =
      value_pack<functions::offset, functions::offset::all,
```
Did you mean this?

vasilytric replied:
yes, applied this comment [here](https://github.com/intel/llvm-test-suite/pull/939/commits/3a3e37f3d8b5c247a3e92f48f51646d4bf27f524#diff-afaea17121d6706af52229dcfb462795dfe86fde5c98f09b836505a741830b6bR228-R231)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,250 @@
+    constexpr DataT base_value = 100;
+    constexpr DataT result_fill_value = 5;
+
+    auto init_values =
+        functions::get_init_values<NumElemsToChange, ChosenOperator, DataT>(
+            base_value);
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> shared_init_values(init_values.begin(),
+                                            init_values.end(), allocator);
```
I was sure you've told that you'd passed allocator to the factory function `get_init_values` itself in https://github.com/intel/llvm-test-suite/pull/939#discussion_r834023451, but I see no any. Is it some some local stash not pushed or do we misunderstand each other maybe?

vasilytric replied:
The `get_offsets` and `get_init_values` function has been renamed into `fill_*` and now they obtain a vector that will be filled, so intermediate variables like `init_values` and `offsets` are useless and they has been deleted
I've done it in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/f79221a335ddfdbd86622f5b4e1d73b1343acb24

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,250 @@
+    // used to calculate expected value for all elements that will be returned
+    // after atomic_update call the sum of changed elements.
+    for (size_t i = 0; i < NumElemsToChange; ++i) {
+      if (filter(i) == 1) {
+        ++num_changed_elems;
+      }
```
```suggestion
      num_changed_elems += (filter(i) == 1);
```

vasilytric replied:
this comment has been applied in this commit: ae69b2dc7d65f231687fb2bdeadb72190d6ef384

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,250 @@
+    }
+
+    for (size_t i = 0; i < NumElems; ++i) {
+      const DataT &retrieved = shared_result[i];
+      if (i < NumberIteractions) {
+        if (result_fill_value != retrieved) {
+          passed = fail_test(i, result_fill_value, retrieved, data_type,
+                             "atomic_update return value", operator_type);
+        }
+      } else {
+        const DataT &expected = base_value * num_changed_elems +
+                                (i - NumberIteractions) * num_changed_elems;
+        if (expected != retrieved) {
+          passed = fail_test(i, expected, retrieved, data_type,
+                             "atomic_update return value", operator_type);
```
Do we really:

- use `atomic_update` over something with `i >= NumberIterations`
- change `shared_result[i]` on device side in case of `i >= NumberIterations`?

Doesn't `shared_result` need only to be `NumberIterations` in size? Or maybe I'm missing something.

vasilytric replied:
This commit has been applied in this commit: 556d3dad42da4b4d9382fb3612f3051931ad6a84
`shared_result` now used `NumberIteractions` as vector size [here](https://github.com/intel/llvm-test-suite/pull/939/commits/556d3dad42da4b4d9382fb3612f3051931ad6a84#diff-afaea17121d6706af52229dcfb462795dfe86fde5c98f09b836505a741830b6bR112-R113)

vasilytric committed:
[SYCL][ESIMD] Use correct sizes for shared data

vasilytric committed:
[SYCL][ESIMD] Fix variable and namespace naming

vasilytric committed:
[SYCL][ESIMD] Return value from kernel function

vasilytric committed:
[SYCL][ESIMD] Remove use of intermediate variable for init data

vasilytric committed:
[SYCL][ESIMD] Disable test due to unexpected error

vasilytric committed:
[SYCL][ESIMD] Remove unused parameter

vasilytric committed:
Remove `if` block while compare num_changed_elems

It reduced number of code strings

vasilytric committed:
[SYCL][ESIMD] Add a semicolon to the end of the statement

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update.hpp:
```diff
@@ -0,0 +1,82 @@
+                "Unsupported operator");
+
+  for (size_t i = 0; i < N; ++i) {
+    vector.push_back(base_value + step * i);
```
Looks like a proper fit for `std::iota`. Are there any architecture concerns why we shouldn't use it?
Still maybe it's better to not have `fill_init_values` at all? See comment below for `functions::fill_init_values` call.

vasilytric replied:
the `fill_init_values` has been deleted [here](https://github.com/intel/llvm-test-suite/pull/939/commits/43ce07f5132c624c1f20ab064dbaf302ce9788e9#diff-28ba9c453e6f6e488b2719b183c72b8802b07162628e6fc1445d3fe6d8206cf6L66-L81)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,239 @@
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> shared_init_values(NumElems, allocator);
+    functions::fill_init_values<NumElems, ChosenOperator, DataT>(
+        base_value, shared_init_values);
```
```suggestion
    std::fill(shared_init_values.begin(), shared_init_values.end(), base_value)
```
By doing so you make explicit statement what we can expect after increment/decrement operators.
So it's easier to read verification code below.

vasilytric replied:
I've used `std::fill` here [here](https://github.com/intel/llvm-test-suite/pull/939/commits/43ce07f5132c624c1f20ab064dbaf302ce9788e9#diff-afaea17121d6706af52229dcfb462795dfe86fde5c98f09b836505a741830b6bL102-R105)

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,239 @@
+    std::sort(shared_result.begin(), shared_result.end());
+    for (size_t i = 0; i < NumberIterations; ++i) {
+      const DataT &retrieved = shared_result[i];
+      const DataT &expected = base_value * num_changed_elems +
+                              (i - NumberIterations) * num_changed_elems;
```
1. As a minor suggestion:

```suggestion
      const DataT &expected = (base_value - NumberIteractions + i) * num_changed_elems;
```
--------------------
2. So that's a sorted sum of values before update. Shouldn't this value be different for `increment` and `decrement` operators? You are starting from `shared_init_values` filled with `base_value` for both.

vasilytric replied:
the new algorithm to calculate expected value for this case that depends on `atomic_op` enum value has been implemented in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/b691d458173b14ad5b2997513cda4d914a364989

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    std::vector<size_t> changed_elems_indexes;
+    // Collect the indexess that has been changed.
```
```suggestion
    // Collect the indexes that has been changed.
```

vasilytric replied:
this comment has been applied in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/6a710603c232774f40423443be7d11b5be6f5013

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,239 @@
+          passed = fail_test(i, expected_after_change, retrieved, data_type,
+                             "value that should be updated", operator_type);
+        }
+        updated_elem_next_index++;
```
Nit, to avoid a temporary creation
```suggestion
        ++updated_elem_next_index;
```

vasilytric replied:
this comment has been applied in this commit: https://github.com/intel/llvm-test-suite/pull/939/commits/6a710603c232774f40423443be7d11b5be6f5013

vasilytric committed:
[SYCL][ESIMD] Change algorithm to verify simd return values

vasilytric committed:
[SYCL][ESIMD] Use preincrement instead of postincrement

vasilytric committed:
[SYCL][ESIMD] Use std::fill instead of functions::fill_init_values

vasilytric committed:
Merge branch 'intel' into atomic_upd_inc_decr

yuriykoch commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,247 @@
+    std::sort(shared_result.begin(), shared_result.end());
+    for (size_t i = 0; i < NumberIterations; ++i) {
+      const DataT &retrieved = shared_result[i];
+      DataT expected;
```
Nit: we might move variable declaration out of loop

vasilytric replied:
The `expected` variable has been moved out of loop in this commit: 91346196a654c6ae28a3a129ed8d2f8d2b9a5e22

vasilytric committed:
[SYCL][ESIMD] Extend type coverage

Use new ways to filter types by they size

vasilytric committed:
[SYCL][ESIMD] Move variable declaration out of loop

vasilytric committed:
[SYCL][ESIMD] Add one more reason to disable test compilation

Unexpected  static_assert obtained when unsigned long data type is used

vasilytric commented:
@sndmitriev Would you mind to review this PR?

kbobrovs commented on SYCL/ESIMD/api/functional/functions/functions_atomic_update_inc_dec.cpp:
```diff
@@ -0,0 +1,266 @@
+      cgh.parallel_for<
+          Kernel<DataT, NumElems, TestCaseT, ChosenOperatorT, NDRangeDimT,
+                 NumElemsToChangeT, ChangeElemFilterT, AlgorithmToChangeT>>(
+          nd_range, [=](sycl::nd_item<1> nd_item) SYCL_ESIMD_KERNEL {
```
The test seems too weak. 16 threads each doing an atomic update is likely to succeed even w/o atomic access. There should be many more threads, ideally, more than there are hardware threads, and each thread should do atomic update in a loop for some time to achieve actual contention.

yuriykoch replied:
To think of, we also might use a barrier to sync the start of work-items execution.

vasilytric replied:
@kbobrovs Is it ok if we will obtain max number of threads by using `sycl::device::get_info<info::device::max_compute_units>()`?

> There should be many more threads, ideally, more than there are hardware threads


Also, number of threads should be less than `sycl::device::get_info<info::device::max_work_group_size>()`, I am right?

The following device information descriptors:
- info::device::max_compute_units
- info::device::max_work_group_size

Described in [SYCL-2020 #4.6.4.2. Device information descriptors](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_device_information_descriptors)

kbobrovs replied:
> Also, number of threads should be less than `sycl::device::get_info<info::device::max_work_group_size>()`, I am right?

`max_work_group_size()` does not reflect the number of h/w threads. It should be a function of `max_compute_units>()` (not sure about the exact formula). But NDRange can definitely exceed `max_hardware_threads * sub_group_size` . S/W threads will be mapped to hardware in "rounds". That is my suggestion.

> To think of, we also might use a barrier to sync the start of work-items execution.

It is always a good idea to reduce the number of features used in a test to make it more focused. I think it will be strong enough w/o barriers if there is a loop in each thread. Besides, we only have group barriers, no machine-wide ones.

kbobrovs committed:
Merge branch 'intel' into atomic_upd_inc_decr
